lib/quicly.c

```diff
@@ -309,6 +343,19 @@ struct st_quicly_conn_t {
         uint8_t pending_flows;
 #define QUICLY_PENDING_FLOW_NEW_TOKEN_BIT (1 << 5)
 #define QUICLY_PENDING_FLOW_HANDSHAKE_DONE_BIT (1 << 6)
+        /**
+         * pending RETIRE_CONNECTION_ID frames to be sent, as a response to retire prior to in NEW_CONNECTION_ID
+         */
+        struct {
+            /**
+             * sequence numbers to ask for retirement -- UINT64_MAX indicates empty slot
+             */
+            uint64_t sequences[QUICLY_RETIRE_CONNECTION_ID_LIMIT];
+            /**
+             * # of active elements in sequences
+             */
+            uint64_t num_pending;
```

Comment:
Instead of having `num_pending`, I kind of wonder if it would be possible to use a bit of `pending_flows` to indicate if there's more than one entry in `sequence`?

That would reduce the number of CPU cycles spent in `quicly_get_first_timeout`, and also free us from the burden of mainitaning `sequences` and `num_pending` in sync.

`sequences` can be a list of valid sequence numbers that is terminated by `UINT64_MAX`. When adding a CID to the list, you would append a CID, then, when sending RCID frames, you would pop one entry from the beginning of the list.

Reply:
👍 Reducing `if` branch in a critical path sounds too attractive to ignore :) Let me try to see if this works.